src/gwt/src/org/rstudio/core/client/virtualscroller/virtualscroller.js

```diff
@@ -0,0 +1,299 @@
+
+(function () {
+  "use strict";
+  VirtualScroller = function () {
+  };
+
+  VirtualScroller.prototype = {
+    setup: function(element) {
+      if (!element)
+        return;
+
+      //  *** _CONSTANTS ***
+      this._DEBUG = true;
+      this._SCROLL_DEBOUNCE_MS = 500;
+      this._BUCKET_MAX_SIZE = 50;
+      this._MAX_VISIBLE_BUCKETS = 10;
+
+      // we use this style to keep the elements in the DOM for screen readers
+      this._HIDDEN_STYLE = "org-rstudio-core-client-theme-res-ThemeStyles-visuallyHidden";
+
+      //  *** global vars ***
+      this.LAST_SCROLL_TIME = 0;
+      this.INITIALIZED = false;
+
+      //  *** instance vars ***
+      this.scrollerEle = null;
+      this.consoleEle = element;
+      this.buckets = [];
+      this.visibleBuckets = [];
+
+      var self = this;
+
+      self._createAndAddNewBucket = self._createAndAddNewBucket.bind(self);
+      self._debug = self._debug.bind(self);
+      self._getCurBucket = self._getCurBucket.bind(self);
+      self._hideBucket = self._hideBucket.bind(self);
+      self._isAtBottomBucket = self._isAtBottomBucket.bind(self);
+      self._isAtTopBucket = self._isAtTopBucket.bind(self);
+      self._isBucketHidden = self._isBucketHidden.bind(self);
+      self._jumpToBottom = self._jumpToBottom.bind(self);
+      self._moveWindow = self._moveWindow.bind(self);
+      self._onParentScroll = self._onParentScroll.bind(self);
+      self._showBucket = self._showBucket.bind(self);
+      self.append = self.append.bind(self);
+      self.clear = self.clear.bind(self);
+      self.setScrollParent = self.setScrollParent.bind(self);
+
+      // set up the initial bucket
+      this.append(document.createElement("span"));
+
+      // iterate through all historical elements and put them into buckets
+      while(element.children.length > 0) {
+        var child = element.children[0];
+        element.removeChild(child);
+        this.append(child);
+      }
+
+      // put the bucket into the console
+      for (var i = 0; i < this.buckets.length; i++)
+        this.consoleEle.appendChild(this.buckets[i]);
+
+      // traverse up the parents until we find the ace_scroller and attach to it
+      var ancestor = element.parentElement;
+      while (!self.scrollerEle && !!ancestor) {
+        if (ancestor.className.indexOf("ace_scroller") !== -1) {
+          self.setScrollParent(ancestor);
+          ancestor = null;
+        } else {
+          ancestor = ancestor.parentElement;
+        }
+      }
+
+      // jump to latest button
+      this.jumpToLatestButton = document.createElement("div");
+      this.jumpToLatestButton.classList.add("jump-to-latest-console");
+      this.jumpToLatestButton.innerText = "Latest";
+      this.jumpToLatestButton.style.display = "none";
+      this.jumpToLatestButton.onclick = function() {
+        self._jumpToBottom();
+        self._setJumpToLatestVisible(false);
+        document.getElementById("rstudio_console_output").focus();
```

Comment:
This would need to be removed to make the virtual scroller usable in other panes. Maybe it should be supplied as a callback?